File: Suspension matlab/Natural_frequency/Maximum_compression/Maximum_compression.m
clc; clear; close all;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end
g = max_g;

%% data
m = max_m;        % 直量kg
ax = max_ax;      % 最大加速度m/s^2
ay = max_ay;
h = max_h;        % 重心高m
L = max_L;        % 軸距m
t = max_t;        % 輪距

%% 計算
% ===== sweep allowable deformation =====
s = max_s; % 容許壓縮

Fx = m * ax * g * h/(L/2);
Fy = m * ay * g * h/(t/2);
% ===== pitch stiffness =====
K_pitch = Fx./s/2;
K_roll = Fy./s/2;
% ===== plot =====
figure;
plot(K_pitch, s, '-r');   % 畫 pitch 曲線
hold on;
plot(K_roll, s, '-b');    % 畫 roll 曲線
yline(0.025, '--w');% 賽規限制
grid on;

legend("pitch", "roll");  % 加上圖例


xlabel('K');
ylabel('懸吊壓縮');
title('Pitch Stiffness Requirement vs Allowable Deformation');

File: Suspension matlab/Natural_frequency/Maximum_compression/Maximum_compression.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    g = p["max_g"]
    m = p["max_m"]
    ax = p["max_ax"]
    ay = p["max_ay"]
    h = p["max_h"]
    length = p["max_L"]
    track = p["max_t"]
    s = np.asarray(p["max_s"], dtype=float)

    fx = m * ax * g * h / (length / 2)
    fy = m * ay * g * h / (track / 2)
    k_pitch = fx / s / 2
    k_roll = fy / s / 2

    plt.figure()
    plt.plot(k_pitch, s, "-r", label="pitch")
    plt.plot(k_roll, s, "-b", label="roll")
    plt.axhline(0.025, linestyle="--", color="k")
    plt.grid(True)
    plt.legend()
    plt.xlabel("K")
    plt.ylabel("Suspension compression")
    plt.title("Pitch Stiffness Requirement vs Allowable Deformation")
    show_plots()


if __name__ == "__main__":
    main()
